File: src/services/health.ts
import { getRedisClient } from "../lib/redis";
import { CircuitBreakerError } from "../errors";

const CIRCUIT_PREFIX = "circuit:";
const HEALTH_PREFIX = "health:";

export type CircuitState = "CLOSED" | "OPEN" | "HALF_OPEN";

export interface HealthMetrics {
  providerId: string;
  circuitState: CircuitState;
  failureCount: number;
  successCount: number;
  lastFailureTime: number | null;
  averageLatency: number | null;
  healthScore: number;
}

export class HealthService {
  private redis = getRedisClient();
  private failureThreshold = 5;
  private recoveryTimeoutMs = 30000;

  async getCircuitState(providerId: string): Promise<CircuitState> {
    const stateKey = `${CIRCUIT_PREFIX}${providerId}:state`;
    const state = await this.redis.get(stateKey);
    return (state as CircuitState) || "CLOSED";
  }

  async canExecute(providerId: string): Promise<boolean> {
    const state = await this.getCircuitState(providerId);

    if (state === "CLOSED") {
      return true;
    }

    if (state === "OPEN") {
      // Check if recovery timeout has passed
      const lastFailureKey = `${CIRCUIT_PREFIX}${providerId}:last_failure`;
      const lastFailure = await this.redis.get(lastFailureKey);

      if (lastFailure) {
        const elapsed = Date.now() - parseInt(lastFailure, 10);
        if (elapsed >= this.recoveryTimeoutMs) {
          // Transition to HALF_OPEN
          await this.setCircuitState(providerId, "HALF_OPEN");
          await this.redis.set(`${CIRCUIT_PREFIX}${providerId}:failures`, "0");
          await this.redis.set(`${CIRCUIT_PREFIX}${providerId}:successes`, "0");
          return true;
        }
      }
      return false;
    }

    if (state === "HALF_OPEN") {
      // Allow limited requests in half-open state
      const successes = parseInt(
        (await this.redis.get(`${CIRCUIT_PREFIX}${providerId}:successes`)) ||
          "0",
        10,
      );
      const failures = parseInt(
        (await this.redis.get(`${CIRCUIT_PREFIX}${providerId}:failures`)) ||
          "0",
        10,
      );
      return successes + failures < 1; // Allow 1 probe request
    }

    return false;
  }

  /**
   * Check circuit breaker and throw CircuitBreakerError if circuit is open
   * Use this when you want to fail fast with a clear error
   */
  async checkCircuitBreakerOrThrow(providerId: string): Promise<void> {
    const state = await this.getCircuitState(providerId);

    if (state === "OPEN") {
      throw new CircuitBreakerError(
        `Circuit breaker is OPEN for provider ${providerId}`,
        providerId,
        "OPEN",
      );
    }

    if (state === "HALF_OPEN") {
      // Check if probe limit reached
      const successes = parseInt(
        (await this.redis.get(`${CIRCUIT_PREFIX}${providerId}:successes`)) ||
          "0",
        10,
      );
      const failures = parseInt(
        (await this.redis.get(`${CIRCUIT_PREFIX}${providerId}:failures`)) ||
          "0",
        10,
      );

      if (successes + failures >= 1) {
        throw new CircuitBreakerError(
          `Circuit breaker is HALF_OPEN for provider ${providerId}, probe limit reached`,
          providerId,
          "HALF_OPEN",
        );
      }
    }
  }

  async recordSuccess(providerId: string, latencyMs: number): Promise<void> {
    const state = await this.getCircuitState(providerId);
    const successesKey = `${CIRCUIT_PREFIX}${providerId}:successes`;
    const failuresKey = `${CIRCUIT_PREFIX}${providerId}:failures`;

    if (state === "HALF_OPEN") {
      const successes = await this.redis.incr(successesKey);
      if (successes >= 1) {
        // Close the circuit
        await this.setCircuitState(providerId, "CLOSED");
        await this.redis.set(failuresKey, "0");
        await this.redis.set(successesKey, "0");
      }
    } else if (state === "CLOSED") {
      // Reset failure count on success
      await this.redis.set(failuresKey, "0");
    }

    // Record latency with 1 hour TTL
    const latencyKey = `${HEALTH_PREFIX}${providerId}:latency`;
    await this.redis.setex(latencyKey, 3600, latencyMs.toString());

    // Update health score
    await this.updateHealthScore(providerId);
  }

  async recordFailure(providerId: string): Promise<void> {
    const state = await this.getCircuitState(providerId);
    const failuresKey = `${CIRCUIT_PREFIX}${providerId}:failures`;
    const lastFailureKey = `${CIRCUIT_PREFIX}${providerId}:last_failure`;

    const failures = await this.redis.incr(failuresKey);
    await this.redis.set(lastFailureKey, Date.now().toString());

    if (state === "HALF_OPEN") {
      // Re-open the circuit
      await this.setCircuitState(providerId, "OPEN");
    } else if (state === "CLOSED" && failures >= this.failureThreshold) {
      // Open the circuit
      await this.setCircuitState(providerId, "OPEN");
    }

    // Update health score
    await this.updateHealthScore(providerId);
  }

  async getHealthMetrics(providerId: string): Promise<HealthMetrics> {
    const [
      circuitState,
      failureCount,
      successCount,
      lastFailureTime,
      averageLatency,
      healthScore,
    ] = await Promise.all([
      this.getCircuitState(providerId),
      this.redis
        .get(`${CIRCUIT_PREFIX}${providerId}:failures`)
        .then((v) => parseInt(v || "0", 10)),
      this.redis
        .get(`${CIRCUIT_PREFIX}${providerId}:successes`)
        .then((v) => parseInt(v || "0", 10)),
      this.redis
        .get(`${CIRCUIT_PREFIX}${providerId}:last_failure`)
        .then((v) => (v ? parseInt(v, 10) : null)),
      this.redis
        .get(`${HEALTH_PREFIX}${providerId}:latency`)
        .then((v) => (v ? parseInt(v, 10) : null)),
      this.redis
        .get(`${HEALTH_PREFIX}${providerId}:score`)
        .then((v) => (v ? parseFloat(v) : 1.0)),
    ]);

    return {
      providerId,
      circuitState,
      failureCount,
      successCount,
      lastFailureTime,
      averageLatency,
      healthScore,
    };
  }

  async getAllHealthMetrics(): Promise<HealthMetrics[]> {
    // Get all provider IDs from circuit keys
    const keys = await this.redis.keys(`${CIRCUIT_PREFIX}*:state`);
    const providerIds = keys.map((k) =>
      k.replace(`${CIRCUIT_PREFIX}`, "").replace(":state", ""),
    );

    const metrics = await Promise.all(
      providerIds.map((id) => this.getHealthMetrics(id)),
    );

    return metrics;
  }

  private async setCircuitState(
    providerId: string,
    state: CircuitState,
  ): Promise<void> {
    const stateKey = `${CIRCUIT_PREFIX}${providerId}:state`;
    await this.redis.set(stateKey, state);
  }

  private async updateHealthScore(providerId: string): Promise<void> {
    const failures = parseInt(
      (await this.redis.get(`${CIRCUIT_PREFIX}${providerId}:failures`)) || "0",
      10,
    );
    const state = await this.getCircuitState(providerId);

    // Simple health score calculation
    let score = 1.0;

    if (state === "OPEN") {
      score = 0;
    } else if (state === "HALF_OPEN") {
      score = 0.5;
    } else {
      // Penalize consecutive failures
      if (failures > 3) {
        score = 0.5;
      } else if (failures > 0) {
        score = 1 - failures * 0.1;
      }
    }

    const scoreKey = `${HEALTH_PREFIX}${providerId}:score`;
    await this.redis.setex(scoreKey, 3600, score.toString());
  }
}

export const healthService = new HealthService();

File: internal/services/health.go
package services

import (
	"context"
	"fmt"
	"strconv"
	"time"

	"github.com/abdo-355/llm-gateway/internal/config"
	"github.com/abdo-355/llm-gateway/internal/errors"
	"github.com/abdo-355/llm-gateway/internal/logger"
	"github.com/abdo-355/llm-gateway/internal/metrics"
	"github.com/redis/go-redis/v9"
)

type CircuitState string

const (
	StateClosed   CircuitState = "CLOSED"
	StateOpen     CircuitState = "OPEN"
	StateHalfOpen CircuitState = "HALF_OPEN"
)

type ProviderModelPair struct {
	ProviderID string
	Model      string
}

type HealthMetrics struct {
	ProviderID      string       `json:"provider_id"`
	Model           string       `json:"model"`
	CircuitState    CircuitState `json:"circuit_state"`
	FailureCount    int          `json:"failure_count"`
	SuccessCount    int          `json:"success_count"`
	LastFailureTime *int64       `json:"last_failure_time,omitempty"`
	AverageLatency  *int         `json:"average_latency_ms,omitempty"`
	HealthScore     float64      `json:"health_score"`
}

type HealthService struct {
	redis            *redis.Client
	prefix           string
	failureThreshold int
	recoveryTimeout  time.Duration
}

func NewHealthService(redisClient *redis.Client, keyPrefix string) *HealthService {
	prefix := keyPrefix
	if prefix == "" {
		prefix = "health"
	}
	return &HealthService{
		redis:            redisClient,
		prefix:           prefix,
		failureThreshold: 3,
		recoveryTimeout:  30 * time.Second,
	}
}

func (s *HealthService) buildCircuitKeyPrefix(providerID, model string) string {
	return fmt.Sprintf("circuit:%s:%s:%s", s.prefix, providerID, model)
}

func (s *HealthService) buildHealthKeyPrefix(providerID, model string) string {
	return fmt.Sprintf("health:%s:%s:%s", s.prefix, providerID, model)
}

func (s *HealthService) GetCircuitState(ctx context.Context, providerID, model string) CircuitState {
	circuitPrefix := s.buildCircuitKeyPrefix(providerID, model)
	stateKey := fmt.Sprintf("%s:state", circuitPrefix)

	state, err := s.redis.Get(ctx, stateKey).Result()
	if err == redis.Nil {
		return StateClosed
	}
	if err != nil {
		logger.Error().
			Str("type", "app").
			Str("event", "health.circuit_state_failed").
			Err(err).
			Msg("Failed to get circuit state")
		return StateClosed
	}

	return CircuitState(state)
}

func (s *HealthService) CanExecute(ctx context.Context, providerID, model string) bool {
	state := s.GetCircuitState(ctx, providerID, model)
	circuitPrefix := s.buildCircuitKeyPrefix(providerID, model)

	switch state {
	case StateClosed:
		return true

	case StateOpen:
		lastFailureKey := fmt.Sprintf("%s:last_failure", circuitPrefix)

		lastFailure, err := s.redis.Get(ctx, lastFailureKey).Int64()
		if err != nil {
			return false
		}

		if time.Since(time.UnixMilli(lastFailure)) >= s.recoveryTimeout {
			s.setCircuitState(ctx, providerID, model, StateHalfOpen)
			s.redis.Set(ctx, fmt.Sprintf("%s:failures", circuitPrefix), 0, 0)
			s.redis.Set(ctx, fmt.Sprintf("%s:successes", circuitPrefix), 0, 0)
			return true
		}
		return false

	case StateHalfOpen:
		// Allow up to 2 probe requests in HALF_OPEN before requiring a decision
		successes, _ := s.redis.Get(ctx, fmt.Sprintf("%s:successes", circuitPrefix)).Int()
		failures, _ := s.redis.Get(ctx, fmt.Sprintf("%s:failures", circuitPrefix)).Int()
		return successes+failures < 2
	}

	return false
}

func (s *HealthService) CheckCircuitBreaker(ctx context.Context, providerID, model string) error {
	if !s.CanExecute(ctx, providerID, model) {
		state := s.GetCircuitState(ctx, providerID, model)
		return errors.NewCircuitBreakerError(
			fmt.Sprintf("Circuit breaker is %s for provider %s model %s", state, providerID, model),
			providerID, string(state),
		)
	}
	return nil
}

func (s *HealthService) RecordSuccess(ctx context.Context, providerID, model string, latencyMs int) {
	state := s.GetCircuitState(ctx, providerID, model)
	circuitPrefix := s.buildCircuitKeyPrefix(providerID, model)
	healthPrefix := s.buildHealthKeyPrefix(providerID, model)

	var failures int
	if state == StateHalfOpen {
		successesKey := fmt.Sprintf("%s:successes", circuitPrefix)
		successes, _ := s.redis.Incr(ctx, successesKey).Result()
		s.redis.Expire(ctx, successesKey, 24*time.Hour)

		if successes >= 2 {
			s.setCircuitState(ctx, providerID, model, StateClosed)
			s.redis.Del(ctx, fmt.Sprintf("%s:failures", circuitPrefix))
			s.redis.Del(ctx, successesKey)
			state = StateClosed
		}
	} else if state == StateClosed {
		failuresKey := fmt.Sprintf("%s:failures", circuitPrefix)
		failures, _ = s.redis.Get(ctx, failuresKey).Int()
		if failures > 0 {
			s.redis.Decr(ctx, failuresKey)
			failures--
		}
	}

	latencyKey := fmt.Sprintf("%s:latencies", healthPrefix)
	now := time.Now()
	member := fmt.Sprintf("%d-%d", now.UnixMilli(), now.Nanosecond())

	score := healthScoreFromState(state, failures)
	scoreKey := fmt.Sprintf("%s:score", healthPrefix)

	pipe := s.redis.Pipeline()
	pipe.ZAdd(ctx, latencyKey, redis.Z{Score: float64(latencyMs), Member: member})
	pipe.Expire(ctx, latencyKey, time.Hour)
	pipe.Set(ctx, scoreKey, score, time.Hour)
	pipe.Exec(ctx)
}

func (s *HealthService) RecordFailure(ctx context.Context, providerID, model string) {
	state := s.GetCircuitState(ctx, providerID, model)
	circuitPrefix := s.buildCircuitKeyPrefix(providerID, model)

	failuresKey := fmt.Sprintf("%s:failures", circuitPrefix)
	lastFailureKey := fmt.Sprintf("%s:last_failure", circuitPrefix)

	failures, _ := s.redis.Incr(ctx, failuresKey).Result()

	if state == StateHalfOpen {
		s.setCircuitState(ctx, providerID, model, StateOpen)
		state = StateOpen
	} else if state == StateClosed && failures >= int64(s.failureThreshold) {
		s.setCircuitState(ctx, providerID, model, StateOpen)
		state = StateOpen
	}

	score := healthScoreFromState(state, int(failures))
	healthPrefix := s.buildHealthKeyPrefix(providerID, model)
	scoreKey := fmt.Sprintf("%s:score", healthPrefix)

	pipe := s.redis.Pipeline()
	pipe.Expire(ctx, failuresKey, 24*time.Hour)
	pipe.Set(ctx, lastFailureKey, time.Now().UnixMilli(), 24*time.Hour)
	pipe.Set(ctx, scoreKey, score, time.Hour)
	pipe.Exec(ctx)
}

func (s *HealthService) GetHealthMetrics(ctx context.Context, providerID, model string) HealthMetrics {
	circuitState := s.GetCircuitState(ctx, providerID, model)
	circuitPrefix := s.buildCircuitKeyPrefix(providerID, model)
	healthPrefix := s.buildHealthKeyPrefix(providerID, model)

	failuresKey := fmt.Sprintf("%s:failures", circuitPrefix)
	successesKey := fmt.Sprintf("%s:successes", circuitPrefix)
	lastFailureKey := fmt.Sprintf("%s:last_failure", circuitPrefix)
	latencyKey := fmt.Sprintf("%s:latencies", healthPrefix)
	scoreKey := fmt.Sprintf("%s:score", healthPrefix)

	pipe := s.redis.Pipeline()
	failuresCmd := pipe.Get(ctx, failuresKey)
	successesCmd := pipe.Get(ctx, successesKey)
	lastFailureCmd := pipe.Get(ctx, lastFailureKey)
	scoreCmd := pipe.Get(ctx, scoreKey)
	// Issue #8: Use ZRangeWithScores to get latency values from scores
	latencyCmd := pipe.ZRangeWithScores(ctx, latencyKey, 0, -1)

	_, err := pipe.Exec(ctx)
	if err != nil && err != redis.Nil {
		logger.Error().
			Str("type", "db").
			Str("event", "health.metrics_failed").
			Err(err).
			Msg("Failed to get health metrics")
	}

	failureCount, _ := strconv.Atoi(failuresCmd.Val())
	successCount, _ := strconv.Atoi(successesCmd.Val())

	var lastFailureTime *int64
	if lastFailureCmd.Val() != "" {
		if ts, err := strconv.ParseInt(lastFailureCmd.Val(), 10, 64); err == nil {
			lastFailureTime = &ts
		}
	}

	var avgLatency *int
	latencyScores := latencyCmd.Val()
	if len(latencyScores) > 0 {
		sum := 0
		for _, z := range latencyScores {
			sum += int(z.Score)
		}
		avg := sum / len(latencyScores)
		avgLatency = &avg
	}

	healthScore := 1.0
	if scoreCmd.Val() != "" {
		if score, err := strconv.ParseFloat(scoreCmd.Val(), 64); err == nil {
			healthScore = score
		}
	}

	return HealthMetrics{
		ProviderID:      providerID,
		Model:           model,
		CircuitState:    circuitState,
		FailureCount:    failureCount,
		SuccessCount:    successCount,
		LastFailureTime: lastFailureTime,
		AverageLatency:  avgLatency,
		HealthScore:     healthScore,
	}
}

func (s *HealthService) GetAllHealthMetrics(ctx context.Context) []HealthMetrics {
	providers := config.GetProviders()

	metrics := make([]HealthMetrics, 0)
	for _, provider := range providers {
		for _, model := range provider.Models.List {
			metrics = append(metrics, s.GetHealthMetrics(ctx, provider.ID, model))
		}
	}

	return metrics
}

func (s *HealthService) setCircuitState(ctx context.Context, providerID, model string, newState CircuitState) {
	circuitPrefix := s.buildCircuitKeyPrefix(providerID, model)
	stateKey := fmt.Sprintf("%s:state", circuitPrefix)

	oldState := s.GetCircuitState(ctx, providerID, model)

	s.redis.Set(ctx, stateKey, string(newState), 24*time.Hour)

	metrics.CircuitBreakerState.WithLabelValues(providerID, model).Set(
		metrics.CircuitStateToFloat64(string(newState)),
	)

	if oldState != newState {
		metrics.CircuitBreakerTransitionsTotal.WithLabelValues(
			providerID, model, string(oldState), string(newState),
		).Inc()
	}
}

func healthScoreFromState(state CircuitState, failures int) float64 {
	switch state {
	case StateOpen:
		return 0.0
	case StateHalfOpen:
		return 0.5
	default:
		if failures > 3 {
			return 0.5
		} else if failures > 0 {
			return 1.0 - float64(failures)*0.1
		}
		return 1.0
	}
}

// BatchCircuitStates reads circuit breaker states for multiple provider/model pairs
// in a single pipeline round trip.
func (s *HealthService) BatchCircuitStates(ctx context.Context, pairs []ProviderModelPair) map[string]CircuitState {
	if len(pairs) == 0 {
		return nil
	}

	pipe := s.redis.Pipeline()
	type cmdEntry struct {
		key  string
		cmd  *redis.StringCmd
		pair ProviderModelPair
	}
	entries := make([]cmdEntry, len(pairs))
	for i, p := range pairs {
		key := fmt.Sprintf("%s:state", s.buildCircuitKeyPrefix(p.ProviderID, p.Model))
		entries[i] = cmdEntry{key: key, cmd: pipe.Get(ctx, key), pair: p}
	}
	pipe.Exec(ctx)

	result := make(map[string]CircuitState, len(pairs))
	for _, e := range entries {
		key := e.pair.ProviderID + "/" + e.pair.Model
		val, err := e.cmd.Result()
		if err != nil || val == "" {
			result[key] = StateClosed
		} else {
			result[key] = CircuitState(val)
		}
	}
	return result
}

// BatchGetHealthMetrics reads health metrics for multiple provider/model pairs
// in a single pipeline round trip.
func (s *HealthService) BatchGetHealthMetrics(ctx context.Context, pairs []ProviderModelPair) []HealthMetrics {
	if len(pairs) == 0 {
		return nil
	}

	pipe := s.redis.Pipeline()
	type cmdGroup struct {
		pair            ProviderModelPair
		stateCmd        *redis.StringCmd
		failuresCmd     *redis.StringCmd
		successesCmd    *redis.StringCmd
		lastFailureCmd  *redis.StringCmd
		scoreCmd        *redis.StringCmd
		latencyCmd      *redis.ZSliceCmd
	}
	groups := make([]cmdGroup, len(pairs))

	for i, p := range pairs {
		circuitPrefix := s.buildCircuitKeyPrefix(p.ProviderID, p.Model)
		healthPrefix := s.buildHealthKeyPrefix(p.ProviderID, p.Model)
		groups[i] = cmdGroup{
			pair:           p,
			stateCmd:       pipe.Get(ctx, fmt.Sprintf("%s:state", circuitPrefix)),
			failuresCmd:    pipe.Get(ctx, fmt.Sprintf("%s:failures", circuitPrefix)),
			successesCmd:   pipe.Get(ctx, fmt.Sprintf("%s:successes", circuitPrefix)),
			lastFailureCmd: pipe.Get(ctx, fmt.Sprintf("%s:last_failure", circuitPrefix)),
			scoreCmd:       pipe.Get(ctx, fmt.Sprintf("%s:score", healthPrefix)),
			latencyCmd:     pipe.ZRangeWithScores(ctx, fmt.Sprintf("%s:latencies", healthPrefix), 0, -1),
		}
	}
	_, _ = pipe.Exec(ctx)

	results := make([]HealthMetrics, len(groups))
	for i, g := range groups {
		failureCount, _ := strconv.Atoi(g.failuresCmd.Val())
		successCount, _ := strconv.Atoi(g.successesCmd.Val())

		circuitState := StateClosed
		if val, err := g.stateCmd.Result(); err == nil && val != "" {
			circuitState = CircuitState(val)
		}

		var lastFailureTime *int64
		if val := g.lastFailureCmd.Val(); val != "" {
			if ts, err := strconv.ParseInt(val, 10, 64); err == nil {
				lastFailureTime = &ts
			}
		}

		var avgLatency *int
		if scores := g.latencyCmd.Val(); len(scores) > 0 {
			sum := 0
			for _, z := range scores {
				sum += int(z.Score)
			}
			avg := sum / len(scores)
			avgLatency = &avg
		}

		healthScore := 1.0
		if val := g.scoreCmd.Val(); val != "" {
			if score, err := strconv.ParseFloat(val, 64); err == nil {
				healthScore = score
			}
		}

		results[i] = HealthMetrics{
			ProviderID:      g.pair.ProviderID,
			Model:           g.pair.Model,
			CircuitState:    circuitState,
			FailureCount:    failureCount,
			SuccessCount:    successCount,
			LastFailureTime: lastFailureTime,
			AverageLatency:  avgLatency,
			HealthScore:     healthScore,
		}
	}
	return results
}
